File: src/shared/BasicCommerce.Domain/Entities/Product.cs
using BasicCommerce.Domain.Enums;
using BasicCommerce.Domain.ValueObjects;

namespace BasicCommerce.Domain.Entities;

public class Product : TenantEntity
{
    public string Sku { get; private set; } = default!;
    public string Barcode { get; private set; } = default!;
    public string? Plu { get; private set; }
    public string Name { get; private set; } = default!;
    public string NameBn { get; private set; } = default!;
    public string? Description { get; private set; }
    public Guid CategoryId { get; private set; }
    public Money Price { get; private set; } = default!;
    public Money? CostPrice { get; private set; }
    public UnitType UnitType { get; private set; }
    public string? UnitLabel { get; private set; }
    public Guid VatRateId { get; private set; }
    public bool IsWeightBased { get; private set; }
    public bool IsAgeRestricted { get; private set; }
    public int? AgeRestrictionYears { get; private set; }
    public bool IsEbtEligible { get; private set; }
    public bool TrackInventory { get; private set; } = true;
    public int ReorderLevel { get; private set; }
    public string? ImageUrl { get; private set; }


    public Category? Category { get; private set; }

    private Product() { }

    public static Product Create(Guid tenantId, string sku, string barcode, string name,
        string nameBn, Guid categoryId, Money price, Guid vatRateId,
        UnitType unitType = UnitType.Each, bool isWeightBased = false)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(sku);
        ArgumentException.ThrowIfNullOrWhiteSpace(barcode);
        ArgumentException.ThrowIfNullOrWhiteSpace(name);

        return new Product
        {
            TenantId = tenantId,
            Sku = sku.ToUpperInvariant(),
            Barcode = barcode,
            Name = name,
            NameBn = nameBn,
            CategoryId = categoryId,
            Price = price,
            VatRateId = vatRateId,
            UnitType = unitType,
            IsWeightBased = isWeightBased
        };
    }

    public void UpdatePrice(Money newPrice)
    {
        Price = newPrice;
        UpdatedAt = DateTime.UtcNow;
    }

    public void UpdateCostPrice(Money? costPrice)
    {
        CostPrice = costPrice;
        UpdatedAt = DateTime.UtcNow;
    }

    public void UpdateDetails(string name, string nameBn, string? description,
        Guid categoryId, Guid vatRateId, UnitType unitType, string? unitLabel,
        bool isWeightBased, bool isEbtEligible, bool trackInventory,
        int reorderLevel, string? imageUrl, Money? costPrice = null)
    {
        Name = name;
        NameBn = nameBn;
        Description = description;
        CategoryId = categoryId;
        VatRateId = vatRateId;
        UnitType = unitType;
        UnitLabel = unitLabel;
        IsWeightBased = isWeightBased;
        IsEbtEligible = isEbtEligible;
        TrackInventory = trackInventory;
        ReorderLevel = reorderLevel;
        ImageUrl = imageUrl;
        if (costPrice is not null) CostPrice = costPrice;
        UpdatedAt = DateTime.UtcNow;
    }

    public void SetAgeRestriction(int years)
    {
        IsAgeRestricted = true;
        AgeRestrictionYears = years;
    }

    public void RemoveAgeRestriction()
    {
        IsAgeRestricted = false;
        AgeRestrictionYears = null;
    }

    public void SetPlu(string plu) => Plu = plu;

    public void Deactivate()
    {
        Status = EntityStatus.Inactive;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Activate()
    {
        Status = EntityStatus.Active;
        UpdatedAt = DateTime.UtcNow;
    }
}

File: src/clients/BasicCommerce.WebAdmin/src/types/product.ts
export type ProductStatus = 'Active' | 'Inactive';

export type ProductTag = {
  id: string;
  name: string;
};

export type ProductImage = {
  id: string;
  productId: string;
  title: string;
  description: string | null;
  url: string;
  isUploaded: boolean;
  sortOrder: number;
  createdAt: string;
  updatedAt: string | null;
};

export type Product = {
  id: string;
  sku: string;
  barcode: string;
  plu: string | null;
  name: string;
  nameBn: string;
  description: string | null;
  categoryId: string;
  categoryName: string;
  categoryStatus: 'Active' | 'Inactive' | 'Deleted';
  price: number;
  currency: string;
  costPrice: number | null;
  vatRateId: string;
  vatRateName: string;
  vatRate: number;
  unitType: string;
  unitLabel: string | null;
  isWeightBased: boolean;
  isPerishable: boolean;
  isAgeRestricted: boolean;
  ageRestrictionYears: number | null;
  isEbtEligible: boolean;
  trackInventory: boolean;
  reorderLevel: number;
  status: ProductStatus;
  imageUrl: string | null;
  manufacturerId: string | null;
  manufacturerName: string | null;
  tags: ProductTag[];
  createdAt: string;
  updatedAt: string | null;
};

export type ProductFormData = {
  sku: string;
  barcode: string;
  plu: string;
  name: string;
  nameBn: string;
  description: string;
  categoryId: string;
  manufacturerId: string;
  price: string;
  costPrice: string;
  vatRateId: string;
  unitType: string;
  unitLabel: string;
  isWeightBased: boolean;
  isPerishable: boolean;
  isAgeRestricted: boolean;
  ageRestrictionYears: string;
  isEbtEligible: boolean;
  trackInventory: boolean;
  reorderLevel: string;
  imageUrl: string;
  tagIds: string[];
};

export type ProductListItem = {
  id: string;
  sku: string;
  name: string;
  nameBn: string;
  categoryName: string;
  price: number;
  currency: string;
  vatRate: number;
  imageUrl: string | null;
  manufacturerName: string | null;
  status: ProductStatus;
};

export type ProductListResponse = {
  items: ProductListItem[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
};

export type ProductListParams = {
  page: number;
  pageSize: number;
  search?: string;
  categoryId?: string;
  includeInactive?: boolean;
};

export type VatRate = {
  id: string;
  name: string;
  code: string;
  rate: number;
  isDefault: boolean;
  status: string;
};

export type SimpleCategory = {
  id: string;
  name: string;
  parentCategoryId: string | null;
  status: 'Active' | 'Inactive';
};

export type SimpleManufacturer = {
  id: string;
  name: string;
};

export type SimpleTag = {
  id: string;
  name: string;
  status: 'Active' | 'Inactive';
};

export const UNIT_TYPES = ['Each', 'Weight', 'Volume', 'Length', 'Pack'] as const;
